realtek: pcs: groundwork for upcoming extensions/fixes

jonasjelonek opened the pull request:
This does some groundwork for upcoming extensions and fixes. It basically grew out of the wishlist outlined https://github.com/openwrt/openwrt/pull/20292#issuecomment-336 and recent work in #20736.

A lot of these desired items require at least one of two things:
1. storing per-SerDes information
2. decouple Reltek SerDes modes from the kernels' phy modes.

To address 1.: Instead of cramming all per-SerDes information in the global PCS control structure, introduce a separate structure for a SerDes. For now, provide a field to store the current SerDes mode, and move the TX/RX polarity information into that structure.

To address 2.: A dedicated enum for SerDes modes is added to be used internally in the driver instead of kernels `PHY_INTERFACE_MODE_*`. In the past, there have already been issues and confusion with modes supported by Realtek SerDes but not by the kernel (HSGMII). And similar issues will come up when USXGMII and XSGMII are going to be implemented. The accomplish that, decouple our modes from those of the kernel and expect to map between them in the phylink_pcs API functions.

As a side-effect of this, a huge amount of function calls is simplified because we do not need to pass both a reference to the control structure and the SerDes number, instead passing a reference to the SerDes instance is enough. Furthermore, the quite differently implemented calculation of `even`, `odd`, `base`, and `neighbor` SerDes is dropped and replaced by dedicated helpers. Those helpers return the correct reference to the "other" SerDes to work with it instead of fiddling with the SerDes number.

Tested on:
- RTL838x: Zyxel GS1900-10HP B1/v2
- RTL930x: Zyxel XGS1210-12
- RTL931x: Zyxel XS1930-10

jonasjelonek committed:
realtek: pcs: add dedicated enum for SerDes modes

The Realtek SerDes mode capabilities do not map 1:1 to the
PHY_INTERFACE_MODE_* modes used in the kernel and passed to the PCS.
For example, some PHY chips use the proprietary XSGMII mode for which
there isn't an equivalent in the kernel, or HSGMII.

In the past, this led to problems and confusion using kernel's XGMII to
handle the XSGMII mode, and needed a downstream patch for HSGMII. They
have been solved/worked around for now, but XSGMII is currently not
implemented at all. And who knows what might come in the future.

To make our life easier, introduce a dedicated internal representation
of SerDes modes which differs from kernel's PHY_INTERFACE_MODE_*. This
allows us to map "external" modes to different internal modes as needed
instead of carrying the PHY_INTERFACE_MODE_* through the whole SerDes
configuration code. The PCS driver needs to map PHY_INTERFACE_MODE_* to
RTPCS_SDS_MODE_* in pcs_config, and the latter should be used as the
only one.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: make use of SerDes struct in SerDes setup

Make use of the previously added SerDes struct in SerDes setup and all
functions in its call path by removing (ctrl, sds_num) being passed to
every function call and instead just pass the reference to the
corresponding SerDes instance.

Various SerDes calculations for even, odd and neighbor are unified by
switching to previously introduced helpers.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: drop PCS creation without SerDes ref

Since the beginning, the PCS driver had the ability to call its
rtpcs_create without a reference to a valid PCS node. A comment in the
code mentions that this is done for RTL838X and its built-in octa-PHY
which is connected directly instead of via a SerDes. Further
explanations are not provided.

Drop this ability and make the rtpcs_create call in the dsa driver
conditional. As the built-in PHY of RTL838X isn't attached to a SerDes,
there is no obvious point of having the PCS driver in that chain. The
ports are marked as internal and have no pcs-handle, thus no phylink_pcs
instance should be created.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: move polarity into SerDes struct

As a first real usage of the new SerDes struct, move the polarity
configuration there. It was previously located in the global rtpcs_ctrl
struct as an array, indexed by SerDes id. Because this is per-SerDes
information, the new SerDes struct is the correct place to live in.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: rtl930x: drop unused and broken function

Drop the unused and broken function rtpcs_930x_sds_clock_wait from the
PCS driver. The proper working variant is already some lines above and
called rtpcs_930x_sds_wait_clock_ready.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul reviewed (changes requested):
Thanks for your hard work.

jonasjelonek committed:
realtek: pcs: use per-variant SerDes count

Use a separate configuration field for the number of SerDes for each
variant of the Realtek Otto family. Add this field to the config
structure, assign it and use it during driver probe. This narrows
possible error cases and is needed for upcoming extensions.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: add separate SerDes struct

Add a separate structure for a SerDes. This is needed to appropriately
store per-SerDes information, which in turn is needed for future work.
Additionally, it's intended to reduce boilerplate and several
inconsistencies.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: add helpers for even, odd, neighbor SerDes

Add dedicated helpers to get references to even, odd and neigbor SerDes
if needed. This should replace the various calculations scattered
throughout the code, providing a unified way to work with adjacent
SerDes.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: assign SerDes reference upon PCS creation

Upon creation of a phylink_pcs instance by calling rtpcs_create, assign
a reference to the corresponding SerDes to the link structure. In the
next step, this should be used everywhere instead of the plain SerDes
number.

Rename the field used to hold the SerDes number from 'sds' to 'sds_num'
and name the new field 'sds' to make clear what is what.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: drop unneeded SerDes number range checks

By using references to pre-initiated SerDes instances instead of plain
SerDes number, there is no need to check for the range anymore in
various places. During driver/pcs init it is ensured that only valid
SerDes will reach the configuration functions.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul reviewed (changes requested):
Some hints to simplify/split this big commit a little more ...

jonasjelonek committed:
realtek: pcs: switch to id from SerDes struct

Drop usage of the to-be-phased-out SerDes id stored in rtpcs_link and
use the reference to the SerDes instance to use the embedded id in
rtpcs_serdes instead.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

jonasjelonek committed:
realtek: pcs: make use of SerDes struct in set_autoneg

Also switch set_autoneg (and related helper rtpcs_sds_modify) to the
SerDes struct instead of the plain SerDes id by using just the reference
to the SerDes instance instead of (ctrl, sds_id) tuple. This completes
the transition.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21146
Signed-off-by: Robert Marko <robimarko@gmail.com>

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -152,7 +152,8 @@ struct rtpcs_ctrl {
 	struct phylink_pcs pcs;
-	int sds;
+	struct rtpcs_serdes *sds;
+	int sds_id;
```
Looks like unneeded duplication of "link->sds->id".

jonasjelonek replied:
reordered and squashed the commits, fixed.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -911,47 +928,44 @@ static void rtpcs_930x_sds_force_mode(struct rtpcs_ctrl *ctrl, int sds,
-		pr_err("%s: invalid SerDes number: %d\n", __func__, sds);
-		return;
-	}
-
```
This should be part of the subsequent drop-commit.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -1015,17 +1029,18 @@ static void rtpcs_930x_sds_tx_config(struct rtpcs_ctrl *ctrl, int sds,
 /* Wait for clock ready, this assumes the SerDes is in XGMII mode
  * timeout is in ms
  */
+/*
```
Also it may not harm. Commenting out rtpcs_930x_sds_clock_wait is not necessary or should be a separate commit.

jonasjelonek replied:
I could've used the temporary SDS macro or comment to whole function. Nevertheless, this function is useless anyway (you added a proper implementation above during your CMU/PLL refactoring). So I added another pre-commit dropping this function.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2667,10 +2692,13 @@ static int rtpcs_931x_setup_serdes(struct rtpcs_ctrl *ctrl, int sds,
 	int chiptype = 0;
 
-	if (sds < 0 || sds > 13)
+	if (sds_num < 0 || sds_num > 13)
```
Not compile tested. There is no sds_num here.

jonasjelonek replied:
just compile tested this as a whole, not every single commit. Anyway, fixed.

plappermaul commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -233,12 +233,12 @@ static int rtpcs_sds_write_bits(struct rtpcs_serdes *sds, int page,
 }
 
-static int rtpcs_sds_modify(struct rtpcs_ctrl *ctrl, int sds, int page, int regnum,
+static int rtpcs_sds_modify(struct rtpcs_serdes *sds, int page, int regnum,
```
Just a friendly reminder. sds_write_bits() and sds_modify() are basically the same functions. One uses a mask, the other high/low bit id. We should later decide which one to keep.

plappermaul reviewed (changes requested):
Reviewing that large ctrl+sds -> sds conversion commit is quite hard. Nevertheless I understand that it basically follows the same pattern and can only split into multiple commits if we make use of some interim functions.

To sum it up: If you take care about my last comments and run a final compile & test this is good to go for me.

Once again. Thanks a lot.

jonasjelonek commented:
Just for the record: I'm not a fan of these big commits either. I still decided to do it this way. Using interims stuff actually confuses me a lot more (I loose the overview much faster).

Nevertheless, addressed all comments, compile-tested and run-tested on Zyxel XGS1210-12

plappermaul reviewed (approved):
So many changes but only a small leap ahead. You earned the medal of a "cleaner".

jonasjelonek commented:
Let's see which medals I can earn next ^^

GoetzGoerisch reviewed (approved):
run-tested: HPE 1920-24G (JG924A)

Thank you!
